network/w5500_network/w5500_network.c

```diff
@@ -0,0 +1,643 @@
+ * @brief Convert socket_address to w5500_socket_address
+ *
+ * @param src - Source socket address (network interface)
+ * @param dst - Destination socket address (W5500 specific)
+ *
+ * @return 0 on success, negative error code otherwise
+*******************************************************************************/
+static int w5500_net_addr_net_to_w5500(const struct socket_address *src,
+				       struct w5500_socket_address *dst)
+{
+	int result;
+
+	if (!src || !src->addr || !dst)
+		return -EINVAL;
+
+	result = sscanf(src->addr, "%hhu.%hhu.%hhu.%hhu",
+			&dst->ip[0], &dst->ip[1], &dst->ip[2], &dst->ip[3]);
+
+	if (result != 4)
+		return -EINVAL;
+
+	dst->port[0] = W5500_BYTE_HIGH(src->port);
+	dst->port[1] = W5500_BYTE_LOW(src->port);
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Convert w5500_socket_address to socket_address
+ *
+ * @param src - Source socket address (W5500 specific)
+ * @param dst - Destination socket address (network interface)
+ *
+ * @return 0 on success, negative error code otherwise
+*******************************************************************************/
+static int w5500_net_addr_w5500_to_net(struct w5500_socket_address *src,
+				       struct socket_address *dst)
+{
+	static char ip_str[16];
+
+	if (!src || !dst)
+		return -EINVAL;
+
+	snprintf(ip_str, sizeof(ip_str), "%d.%d.%d.%d",
+		 src->ip[0], src->ip[1], src->ip[2], src->ip[3]);
+
+	dst->addr = ip_str;
+	dst->port = src->port[0] << 8 | src->port[1];
+
+	return 0;
+}
+
+/***************************************************************************//**
+ * @brief Open a socket through the network interface
+ *
+ * @param net       - The network interface
+ * @param sock_id   - Pointer to store the assigned socket ID
+ * @param proto     - Protocol to use (TCP, UDP)
+ * @param buff_size - Buffer size for the socket
+ *
+ * @return 0 in case of success, negative error code otherwise
+*******************************************************************************/
+static int32_t w5500_net_socket_open(void *net, uint32_t *sock_id,
+				     enum socket_protocol proto,
+				     uint32_t buff_size)
+{
+	struct w5500_network_dev *net_dev = net;
+	uint8_t buff_kb;
+	uint8_t physical_id;
+	struct w5500_socket_map *socket;
+	uint8_t w5500_protocol;
+	int ret;
+
+	if (buff_size <= 1024)
+		buff_kb = 1;
+	else if (buff_size <= 2048)
+		buff_kb = 2;
+	else if (buff_size <= 4096)
+		buff_kb = 4;
+	else if (buff_size <= 8192)
+		buff_kb = 8;
+	else
+		buff_kb = 16;
+
+	ret = w5500_net_find_free_socket(net_dev, &physical_id);
+	if (ret)
+		return ret;
+
+	w5500_protocol = w5500_net_translate_protocol(proto);
+
+	ret = w5500_socket_open(net_dev->mac_dev, physical_id, w5500_protocol, buff_kb);
+	if (ret)
+		return ret;
+
+	socket = &net_dev->sockets[physical_id];
+
+	socket->role = (proto == PROTOCOL_TCP) ? W5500_ROLE_CLIENT : W5500_ROLE_UNUSED;
```

Comment:
I don't think you can know if the socket will be used by a client or server at this point